syntax/grammar.mjs

```diff
@@ -0,0 +1,502 @@
+        maybe(inline_space),
+        either(
+            sequence(
+                Pattern.abstract,
+                repeat(Attribute).abstract),
+            sequence(
+                always(null).abstract,
+                repeat1(Attribute).abstract)),
+        line_end)
+    .map(flatten(1))
+    .map(keep_abstract)
+    .map(list_into(FTL.Message)));
+
+let Term = defer(() =>
+    sequence(
+        maybe(Comment).abstract,
+        TermIdentifier.abstract,
+        maybe(inline_space),
+        char("="),
+        maybe(inline_space),
+        Pattern.abstract,
+        repeat(Attribute).abstract,
+        line_end)
+    .map(keep_abstract)
+    .map(list_into(FTL.Term)));
+
+let Comment = defer(() =>
+    repeat1(
+        sequence(
+            char("#"),
+            comment_line.abstract))
+    .map(flatten(1))
+    .map(keep_abstract)
+    .map(join)
+    .map(into(FTL.Comment)));
+
+let GroupComment = defer(() =>
+    repeat1(
+        sequence(
+            string("##"),
+            comment_line.abstract))
+    .map(flatten(1))
+    .map(keep_abstract)
+    .map(join)
+    .map(into(FTL.GroupComment)));
+
+let ResourceComment = defer(() =>
+    repeat1(
+        sequence(
+            string("###"),
+            comment_line.abstract))
+    .map(flatten(1))
+    .map(keep_abstract)
+    .map(join)
+    .map(into(FTL.ResourceComment)));
+
+/* ----------------------------------------------------------------- */
+/* Adjacent junk_lines should be joined into FTL.Junk during the AST
+   construction. */
+let junk_line = defer(() =>
+    sequence(
+        regex(/.*/),
+        line_end)
+    .map(join)
+    .map(into(FTL.Junk)));
+
+/* --------------------------------- */
+/* Attributes of Messages and Terms. */
+let Attribute = defer(() =>
+    sequence(
+        break_indent,
+        char("."),
+        Identifier.abstract,
+        maybe(inline_space),
+        char("="),
+        maybe(inline_space),
+        Pattern.abstract)
+    .map(keep_abstract)
+    .map(list_into(FTL.Attribute)));
+
+/* ----------------------------------------------- */
+/* Patterns consist of TextElements or Placeables. */
+let Pattern = defer(() =>
+    repeat1(
+        PatternElement)
+    // Flatten indented Placeables.
+    .map(flatten(1))
+    .map(list_into(FTL.Pattern)));
+
+let PatternElement = defer(() =>
+    either(
+        TextElement,
+        Placeable,
+        sequence(
+            // Trimmed or joined into a preceding FTL.TextElement during the
+            // AST construction.
+            break_indent.map(into(FTL.TextElement)),
+            Placeable)));
+
+let TextElement = defer(() =>
+    repeat1(
+        either(
+            text_char,
+            text_cont))
+    .map(join)
+    .map(into(FTL.TextElement)));
+
+let Placeable = defer(() =>
+    sequence(
+        char("{"),
+        maybe(inline_space),
+        either(
+            // Order matters!
+            BlockExpression,
+            InlineExpression),
```

Comment:
following up on an earlier comment, should this be `either(BE, IE).abstract` and s/element_at/keep_abstract/ ?

Reply:
sounds good. I had the feeling you'd like to keep the maybe() top-level ;-), the way it renders in the ebnf is a good supportive argument

Reply:
Is this comment about the convention or is there some other reason to prefer `keep_abstract`? The reason why I used `element_at` here is that it returns a scalar value so I can then use `into` rather than `list_into` which made a little bit more sense to me semantically. There's only one thing that we care about inside of `Placeable`. I'm not overly attached to this reasoning, however, so I'm happy to change it if you feel it will improve the readability.